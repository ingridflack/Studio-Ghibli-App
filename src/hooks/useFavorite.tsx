import {
  createContext,
  ReactNode,
  useContext,
  useEffect,
  useState,
} from "react";
import { toast } from "react-toastify";
import { IMovie } from "../types";

interface FavoriteProviderProps {
  children: ReactNode;
}

interface FavoriteContextData {
  favorites?: IMovie[];
  checkFavorite: (movie: IMovie) => boolean;
  toggleFavorite: (movie: IMovie) => void;
}

const FavoriteContext = createContext<FavoriteContextData>(
  {} as FavoriteContextData
);

const STORAGE_FAVORITE_KEY = "@Ghibli:favorites";

export function FavoriteProvider({
  children,
}: FavoriteProviderProps): JSX.Element {
  const [favorites, setFavorites] = useState<IMovie[] | undefined>();

  useEffect(() => {
    const storagedFavorites = localStorage.getItem(STORAGE_FAVORITE_KEY);
    const favorites = storagedFavorites ? JSON.parse(storagedFavorites) : [];

    setFavorites(favorites);
  }, []);

  useEffect(() => {
    if (favorites) {
      localStorage.setItem(STORAGE_FAVORITE_KEY, JSON.stringify(favorites));
    }
  }, [favorites]);

  const checkFavorite = (movie: IMovie): boolean =>
    !!favorites?.some((item) => item.id === movie.id);

  const addFavorite = (movie: IMovie) => {
    setFavorites((favorites) => (favorites ? [...favorites, movie] : []));

    toast.success("Movie added to favorites.", {
      position: toast.POSITION.BOTTOM_RIGHT,
    });
  };

  const removeFavorite = (movie: IMovie) => {
    setFavorites((favorites) =>
      favorites?.filter((item) => item.id !== movie.id)
    );

    toast.success("Movie removed from favorites.", {
      position: toast.POSITION.BOTTOM_RIGHT,
    });
  };

  const toggleFavorite = (movie: IMovie) => {
    if (checkFavorite(movie)) {
      removeFavorite(movie);
    } else {
      addFavorite(movie);
    }
  };

  return (
    <FavoriteContext.Provider
      value={{
        favorites,
        checkFavorite,
        toggleFavorite,
      }}
    >
      {children}
    </FavoriteContext.Provider>
  );
}

export function useFavorite() {
  const context = useContext(FavoriteContext);

  return context;
}
