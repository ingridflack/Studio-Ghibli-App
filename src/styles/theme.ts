const theme = {
  colors: {
    background: "#000",

    gray: {
      white: "#f6f6f6",
      platinum: "#e1e1e6",
      silver: "#a8a8b3",
      lightGray: "#7b8c98",
      dark: "#141414",
    },

    primary: "#e50914",
    secondary: "#b20710",
  },
  fontSize: {
    tiny: "0.65rem",
    small: "0.75rem",
    medium: "0.875rem",
    normal: "1rem",
    large: "1.5rem",
  },
} as const;

export type Theme = typeof theme;

export default theme;
