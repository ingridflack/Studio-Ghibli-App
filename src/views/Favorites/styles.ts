import styled, { css } from "styled-components";

export const Title = styled.h2`
  ${({ theme }) => css`
    color: ${theme.colors.gray.white};
    font-weight: 400;
    font-size: 1.4rem;
  `}
`;
