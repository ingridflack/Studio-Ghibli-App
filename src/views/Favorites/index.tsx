import Alert from "../../components/Alert";
import Header from "../../components/Header";
import MovieCard from "../../components/MovieCard";
import { useFavorite } from "../../hooks/useFavorite";
import { IMovie } from "../../types";
import { MoviesList } from "../Home/styles";

import * as S from "./styles";

const Favorites = () => {
  const { favorites, toggleFavorite } = useFavorite();
  return (
    <>
      <Header />

      <S.Title>Favorites</S.Title>

      <MoviesList>
        {favorites?.map((movie: IMovie) => (
          <MovieCard key={movie.id} movie={movie} />
        ))}

        {!favorites?.length && <Alert>No favorites added</Alert>}
      </MoviesList>
    </>
  );
};

export default Favorites;
