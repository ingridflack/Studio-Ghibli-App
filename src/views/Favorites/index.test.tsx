import { render } from "@testing-library/react";
import { ThemeProvider } from "styled-components";
import Favorites from ".";
import { FavoriteProvider } from "../../hooks/useFavorite";
import theme from "../../styles/theme";

const useRouter = jest.spyOn(require("next/router"), "useRouter");
const useFavorite = jest.spyOn(
  require("../../hooks/useFavorite"),
  "useFavorite"
);

const MOVIE_LIST_MOCK = [
  {
    id: "234",
    title: "Title test",
    description: "Description test",
    director: "Director test",
    producer: "Producer test",
    movie_banner: "/assets/studio-ghibli-logo.webp",
    rt_score: "90",
    release_date: "1980",
    people: ["Person 1"],
    original_title: "Original title test",
  },
  {
    id: "2345",
    title: "Title test 2",
    description: "Description test 2",
    director: "Director test 2",
    producer: "Producer test 2",
    movie_banner: "/assets/studio-ghibli-logo.webp",
    rt_score: "90 2",
    release_date: "1980 2",
    people: ["Person 12"],
    original_title: "Original title test 2",
  },
];

describe("Favorites", () => {
  useRouter.mockImplementation(() => ({
    pathname: "/favorites",
  }));

  it("should render the favorites list correctly", () => {
    (useFavorite as jest.Mock).mockReturnValue({
      favorites: MOVIE_LIST_MOCK,
      checkFavorite: jest.fn().mockReturnValue(true),
    });

    const { getAllByTestId } = render(
      <ThemeProvider theme={theme}>
        <FavoriteProvider>
          <Favorites />
        </FavoriteProvider>
      </ThemeProvider>
    );

    const movieCard = getAllByTestId("movie-card-container");

    expect(movieCard).toHaveLength(2);
  });

  it("should render the Alert when the favorites list is empty", () => {
    (useFavorite as jest.Mock).mockReturnValue({
      favorites: [],
      checkFavorite: jest.fn().mockReturnValue(true),
    });

    const { getByTestId } = render(
      <ThemeProvider theme={theme}>
        <FavoriteProvider>
          <Favorites />
        </FavoriteProvider>
      </ThemeProvider>
    );

    const alertContainer = getByTestId("alert-container");

    expect(alertContainer).toBeTruthy();
  });

  it("should match snapshot", () => {
    (useFavorite as jest.Mock).mockReturnValue({
      favorites: MOVIE_LIST_MOCK,
      checkFavorite: jest.fn().mockReturnValue(true),
    });

    const { container } = render(
      <ThemeProvider theme={theme}>
        <FavoriteProvider>
          <Favorites />
        </FavoriteProvider>
      </ThemeProvider>
    );

    expect(container).toMatchSnapshot();
  });
});
