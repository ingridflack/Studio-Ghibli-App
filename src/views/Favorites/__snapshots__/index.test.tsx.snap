// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Favorites should match snapshot 1`] = `
<div>
  <div
    class="sc-c8ddd29-0 iFqajX"
  >
    <img
      alt="Studio Ghibli logo"
      class="sc-c8ddd29-1 bMwEkK"
      height="72"
      src="/assets/studio-ghibli-logo.webp"
      width="128"
    />
    <nav>
      <a
        class="sc-c8ddd29-2 fflUWz"
        href="/"
      >
        Home
      </a>
      <a
        class="sc-c8ddd29-2 etVLLE"
        href="/favorites"
      >
        Favorites
      </a>
    </nav>
  </div>
  <h2
    class="sc-450df110-0 gFJcJn"
  >
    Favorites
  </h2>
  <div
    class="sc-29dbae42-0 jReOdp"
  >
    <a
      class="sc-8c092beb-0 cDXLon"
      data-testid="movie-card-container"
      href="/movies/234"
    >
      <img
        alt="Title test movie banner"
        class="sc-8c092beb-1 iJlXHw"
        data-testid="movie-banner"
        src="/assets/studio-ghibli-logo.webp"
      />
      <div
        class="sc-8c092beb-2 gVYvVF"
      >
        <span
          class="sc-8c092beb-3 iXvlkX"
        >
          Title test
          <div
            class="sc-ade4b8be-1 kbXxyf"
          >
            <div
              class="sc-ade4b8be-0 jUqyaW"
            >
              Remove favorite
            </div>
            <button
              class="sc-8c092beb-5 cREqlB"
              data-testid="favorite-btn"
              type="button"
            >
              <svg
                data-testid="favorited"
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 1024 1024"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M923 283.6a260.04 260.04 0 0 0-56.9-82.8 264.4 264.4 0 0 0-84-55.5A265.34 265.34 0 0 0 679.7 125c-49.3 0-97.4 13.5-139.2 39-10 6.1-19.5 12.8-28.5 20.1-9-7.3-18.5-14-28.5-20.1-41.8-25.5-89.9-39-139.2-39-35.5 0-69.9 6.8-102.4 20.3-31.4 13-59.7 31.7-84 55.5a258.44 258.44 0 0 0-56.9 82.8c-13.9 32.3-21 66.6-21 101.9 0 33.3 6.8 68 20.3 103.3 11.3 29.5 27.5 60.1 48.2 91 32.8 48.9 77.9 99.9 133.9 151.6 92.8 85.7 184.7 144.9 188.6 147.3l23.7 15.2c10.5 6.7 24 6.7 34.5 0l23.7-15.2c3.9-2.5 95.7-61.6 188.6-147.3 56-51.7 101.1-102.7 133.9-151.6 20.7-30.9 37-61.5 48.2-91 13.5-35.3 20.3-70 20.3-103.3.1-35.3-7-69.6-20.9-101.9z"
                />
              </svg>
            </button>
          </div>
        </span>
        <span
          class="sc-8c092beb-4 gAARqp"
        >
          Original title test
        </span>
        <span
          class="sc-8c092beb-6 dawAhA"
        >
          <strong>
            Director:
          </strong>
           
          Director test
        </span>
        <p
          class="sc-8c092beb-7 rPwoM"
        >
          Description test
        </p>
      </div>
    </a>
    <a
      class="sc-8c092beb-0 cDXLon"
      data-testid="movie-card-container"
      href="/movies/2345"
    >
      <img
        alt="Title test 2 movie banner"
        class="sc-8c092beb-1 iJlXHw"
        data-testid="movie-banner"
        src="/assets/studio-ghibli-logo.webp"
      />
      <div
        class="sc-8c092beb-2 gVYvVF"
      >
        <span
          class="sc-8c092beb-3 iXvlkX"
        >
          Title test 2
          <div
            class="sc-ade4b8be-1 kbXxyf"
          >
            <div
              class="sc-ade4b8be-0 jUqyaW"
            >
              Remove favorite
            </div>
            <button
              class="sc-8c092beb-5 cREqlB"
              data-testid="favorite-btn"
              type="button"
            >
              <svg
                data-testid="favorited"
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 1024 1024"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M923 283.6a260.04 260.04 0 0 0-56.9-82.8 264.4 264.4 0 0 0-84-55.5A265.34 265.34 0 0 0 679.7 125c-49.3 0-97.4 13.5-139.2 39-10 6.1-19.5 12.8-28.5 20.1-9-7.3-18.5-14-28.5-20.1-41.8-25.5-89.9-39-139.2-39-35.5 0-69.9 6.8-102.4 20.3-31.4 13-59.7 31.7-84 55.5a258.44 258.44 0 0 0-56.9 82.8c-13.9 32.3-21 66.6-21 101.9 0 33.3 6.8 68 20.3 103.3 11.3 29.5 27.5 60.1 48.2 91 32.8 48.9 77.9 99.9 133.9 151.6 92.8 85.7 184.7 144.9 188.6 147.3l23.7 15.2c10.5 6.7 24 6.7 34.5 0l23.7-15.2c3.9-2.5 95.7-61.6 188.6-147.3 56-51.7 101.1-102.7 133.9-151.6 20.7-30.9 37-61.5 48.2-91 13.5-35.3 20.3-70 20.3-103.3.1-35.3-7-69.6-20.9-101.9z"
                />
              </svg>
            </button>
          </div>
        </span>
        <span
          class="sc-8c092beb-4 gAARqp"
        >
          Original title test 2
        </span>
        <span
          class="sc-8c092beb-6 dawAhA"
        >
          <strong>
            Director:
          </strong>
           
          Director test 2
        </span>
        <p
          class="sc-8c092beb-7 rPwoM"
        >
          Description test 2
        </p>
      </div>
    </a>
  </div>
</div>
`;
